mlir/tools/mlir-cpu-runner/mlir-cpu-runner.cpp

```diff
@@ -17,10 +17,65 @@
 #include "mlir/ExecutionEngine/OptUtils.h"
 #include "mlir/IR/Dialect.h"
 #include "mlir/Target/LLVMIR/Dialect/All.h"
+#include "mlir/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.h"
+#include "mlir/Target/LLVMIR/Export.h"
 
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Linker/Linker.h"
+#include "llvm/Support/CommandLine.h"
 #include "llvm/Support/InitLLVM.h"
 #include "llvm/Support/TargetSelect.h"
 
+using namespace mlir;
+
+llvm::cl::opt<bool> LinkNestedModules(
+    "link-nested-modules",
+    llvm::cl::desc("Link two nested MLIR modules into a single LLVM IR module. "
+                   "Useful if both the host and device code can be run on the "
+                   "same CPU, as in mlir-spirv-cpu-runner tests."));
```

Comment:
Isn't this something that could be done as a mlir-opt pass ahead of time instead of bundling "transformation logic" in the CPU runner?

Reply:
+1

Reply:
Maybe theoretically? But the problem is that mlir-opt produces MLIR, and this linking is done at the LLVM IR level, not at the MLIR one. So we would have to move the conversion to LLVM IR into mlir-opt, then the linking, then it would have to turn that back into MLIR of some kind. I'm not sure what the benefit would be. It's a lot more straightforward to leave this in the runner.

Reply:
Oh I see what you mean, sorry I overlooked the linking logic part.

This flow isn't super clear to me actually, is there still any spirv involved?
Shouldn't we use the same target offloading mechanism in all cases?

Reply:
Yes that's a possibility, we could serialise the device code as LLVM bitcode the same way we do for e.g. SPIR-V. It's probably an unnecessary complication though. I'm willing to something like this out but maybe in a follow-up commit, as there's a few possibilities for how we could proceed?

Reply:
Hmm, I feel like I'm pulled in two different directions here. My goal was to get rid of the separate SPIR-V CPU runner, and the most straightforward way to accomplish that has been to move the codegen steps (SPIR-V -> LLVM MLIR and LLVM MLIR -> LLVM IR) into mlir-opt and mlir-cpu-runner respectively, and that _is_ bringing it more into alignment with the other runners. But for the Vulkan SPIR-V runner, these codegen steps necessarily happen inside the runner, or rather something like them happens inside the Vulkan driver that'll be dynamically linked in (SPIR-V -> [something proprietary] -> LLVM IR -> [native binary]). So if we want the SPIR-V CPU runner to work the same as the Vulkan SPIR-V runner, then we would probably have to move the codegen back into the runner, so that the mlir-opt output can just be SPIR-V.

Reply:
(It's relevant because this is basically just a practice run for getting rid of the Vulkan SPIR-V runner.)

Reply:
We have a similar tug-of-war between our `opt` and `runner` apps (based on the same principles). 

We had to solve the problem of ingress: which IR stage (dialects, passes) to accept. `opt` takes in anything, but `runner` is more complicated, because it needs to lower to LLVM/SPIRV to actually execute, and if the IR is in `linalg`, then it's not going to work. The solution for this is to pass it through `opt` until it's in the right shape, then pipe to `runner`, but defining what the "right shape" is, is the point here.

Let's assume `runner` can only handle LLVM and SPIRV, then the main problem with merging the runners is to document which of these two can be used for which "back-end", and how to drive the back-end (tools, LLVM codegen, dispatch, JIT) with these dialects.

Regardless, I think this is super important and needs to happen. The main difference in code is which MLIR/LLVM functions are called, and some configuration issues, which we can probable common up for all these cases.

Reply:
For my part I think this commit is still an improvement as it stands and I'd rather it be merged more or less as-is (modulo some naming changes etc), then further changes to the shape of this can be experimented with later if necessary. The SPIR-V CPU runner is pretty obscure and rarely used though, so it's not that important that it have the ideal shape, if that makes sense. If there's something worth arguing about it might be how the Vulkan SPIR-V runner changes look.

Reply:
Like, a TODO comment? I guess I can put it in this file next to the flag/pass?

Reply:
Yes, and this TODO comment can link to a github issue

Reply:
That's fine as a stepping stone, thanks again @andfau-amd for the work here.

@fabianmcg : thought about how to make this converge with the rest of the offloading infra? (or maybe it's just not a good fit?)

Reply:
I'll take a closer look tomorrow, but at first glance, it should be possible to generalize and extract the infra from GPU and have something like a general `module-to-offload` for handling nested module compilation. The question is whether people are interested in such a thing outside GPU.

Also, as a reminder for GPUs, we already have integration https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/GPU/SYCL/gpu-reluf32-to-spirv.mlir

Reply:
@fabianmcg keep in mind that while this is a spirv runner, the serialization format for the device code is also llvm. Hence the link step.